### ISSUE TITLE
precedence table

### ISSUE THREAD

**Issue by PavelVozenilek** [andrewrk/zig#114]
Look at the cryptic table:

    x() x[] x.y
    !x -x ~x *x &x ?x %x %%x
    x{}
    * / %
    + - ++
    << >>
    &
    ^
    |
    == != < > <= >=
    &&
    ||
    ?? %%
    = *= /= %= += -= <<= >>= &= ^= |= &&= ||=

Why it is all needed? Who will find the time to learn it? I had not internalized C precedence after two decades. 

Language Pony does it [differently](http://tutorial.ponylang.org/expressions/infix-ops/#precedence):

> Pony takes a different approach and outlaws infix precedence. Any expression where more than one infix operator is used must use parentheses to remove the ambiguity. If you fail to do this the compiler will complain.

**Comment by andrewrk** [comment]
Interesting idea. This is on the table since it adheres to the principle that being readable is more important than being writable.

**Comment by emanuelpalm** [comment]
:+1:

**Comment by andrewrk** [comment]
Some precedence is necessary, such as assignment, function call invocation, array index. Too much parens and the language turns into lisp.

**Comment by PavelVozenilek** [comment]
Assignment, function call, array index are not infix operators. Could there be any ambiguity with them?

I can imagine having default precedence for basic arithmetic (+-/*), this is what little kids get drilled in schools, but IMO it is not worth the effort.

**Comment by PavelVozenilek** [comment]
Also, this feature would make adding user defined operators into the language easier.

**Comment by thejoshwolfe** [comment]
I think this would be a cool idea for some categories of operators, specifically the ones i can never remember the precedence for. e.g. `a && b || c` would be a compile error, or `a & b | c`. i don't think we need to go so far as to outlaw `a + b * c` though.

**Comment by thejoshwolfe** [comment]
see also #29

**Comment by phase** [comment]
My favorite quote from the Pony page is

> Most people will remember to do multiplication before addition, but what about left bit shifting versus bitwise and? Sometimes people misremember (or guess wrong) and that leads to bugs. Worse, those bugs are often very hard to spot.

I love the idea for specific parens. It makes the execution crystal clear.

**Comment by thejoshwolfe** [comment]
The current operator precedence table is inspired by C. But I think Zig can do better than that. For example, `&` > `^` > `|` seems pretty silly. Even if you think `&` > `|` makes sense (because of the analogy to `*` and `+`), it's still pretty confusing that `^` is in the middle of them.

I wanted to write up a proposal here to outline a new precedence table for operators in Zig, but Zig's grammar is a bit more complicated than the current table in `doc/langref.md` shows, and it's not as simple as rearranging and combining some rows.

At the very least, I'd like to see required parentheses for: `&`, `^`, `|`, `<<`, `>>` intermixed with each other, and for `and` and `or` intermixed with each other. There's more to the proposal, but I need to do a more careful analysis of Zig's grammar before posting something comprehensive.

**Comment by thejoshwolfe** [comment]
Here's the real story on Zig's expression grammar (this does not include grammars such as statements, case declarations, top level declarations, etc.). I'll write a follow up comment documenting what i think the expression grammar should be changed to.

There are three types of operator:
* infix operators: always start with `x` and end with `y`, where `x` and `y` are some expressions
* prefix operators: always end with `x`, where `x` is some expression
* suffix operators: always start with `x`, where `x` is some expression
* primary expressions: not really operators, because they are either a single token, or they start and end with unambiguous grouping tokens.

A chainable operator is one that you can use like `a + b + c`, `!!x`, or `a(b)(c)`, and has a well-defined associativity. All infix chainable operators are left-to-right associative (e.g. `a - b - c == (a - b) - c`). Chainable prefix and suffix operators have an obvious associativity (right-to-left and left-to-right respectively). A not-chainable operator cannot be used repeatedly, e.g. `inline inline a`, `a{}{}`, and `a = b = c` are all not allowed.

The following is a list of operator groups. The groups are listed here in order of highest-to-lowest precedence. Each item in a group has equal precedence. If a group is chainable, the operators in the group can be intermixed while chaining (e.g. `a.b[c]`).

primary expressions:
* `123` - and variations, where `123` is a number literal
* `"..."` - where `...` is the characters of the string literal
* `'...'` - where `...` is the characters of the char literal
* `true`, `false`, `null`, `continue`, `undefined`, `error`, `this`, `unreachable`
* `id` - where `id` is an identifier
* `@id(...)` - where `id` is an identifier and `...` is a parameter list
* `goto label` - where `label` is a label
* `break`, `return` - without an expression
* `continue`
* `(x)`
* `{...}` - where `...` is a statement list
* `asm ("...")` - and variations, where `...` is inline assembly
* `struct { ... }` - and variations, where `...` is a container member list
* `if (a) {...}` - and variations, where `...` is a statement list[1]
* `if (a) b else {...}` - and variations, where `...` is a statement list[1]
* `while (a) {...}` - and variations, where `...` is a statement list[1]
* `while (a) b else {...}` - and variations, where `...` is a statement list[1]
* `for (a) {...}` - and variations, where `...` is a statement list[1]
* `for (a) b else {...}` - and variations, where `...` is a statement list
* `switch (a) {...}` - where `...` is a case list
* `comptime {...}` - where `...` is a statement list[1]

prefix, not chainable:
* `inline x`

suffix, chainable:
* `x(...)` - where `...` is a parameter list
* `x[a]` - and variations
* `x.id` - where `id` is an identifier

prefix, chainable:
* `!x`
* `-x`, `-%x`
* `~x`
* `*x`
* `&x` - and variations
* `??x`
* `%%x`
* `?x`
* `%x`
* `[]x` - and variations
* `extern fn() -> x` - and variations

suffix, not chainable:
* `x{...}` - where `...` is a container init body

infix, chainable:
* `x * y`, `x *% y`
* `x / y`
* `x % y`
* `x ** y`

infix, chainable:
* `x + y`, `x +% y`
* `x - y`, `x -% y`
* `x ++ y`

infix, chainable:
* `x << y`, `x <<% y`
* `x >> y`

infix, chainable:
* `x & y`

infix, chainable:
* `x ^ y`

infix, chainable:
* `x | y`

infix, not chainable:
* `x == y` - and variations
* `x < y` - and variations

infix, chainable:
* `x and y`

infix, chainable:
* `x or y`

infix, chainable:
* `x ?? y`
* `x %% y` - and variations

infix, not chainable:
* `x = y` - and variations

prefix, chainable:
* `return x`
* `break x`
* `if (a) x` - and variations[1]
* `if (a) b else x` - and variations[1]
* `while (a) x` - and variations[1]
* `while (a) b else x` - and variations[1]
* `for (a) x` - and variations[1]
* `for (a) b else x` - and variations[1]
* `comptime x`[1]

[1]: for the statement-like expressions `if`, `for`, `while`, and `comptime`, a special exception is made that if a body expression starts with a block (starts with a `{`), then the body expression ends at the end of that block (at the `}`). This disallows, for example, `if (a) {b} + 1 else {c}`.

**Comment by andrewrk** [comment]
For what it's worth I just removed the `inline` prefix operator. (See #306)

**Comment by thejoshwolfe** [comment]
<p>Here's my proposal for the "operator precedence table". It's a lot more complicated than a table.</p>

<p>Before people automatically reject this grammar for being overly complicated, the goal here is to make Zig source code less complicated. My goal for grouping everything the way I did was so that any time a user would ask "Which of these two operators has precedence over the other.", the answer is that they're incompatible, and you need to use parentheses.</p>

<p>Now clearly it's difficult to predict when users will ask that, and so most of the proposal is what I think is a good idea to me, and it's wildly up for debate and discussion. For example, why did I make modulo not heterogeneously chainable with multiplication? Because it seemed like you wouldn't really need that, and it would be more clear with parentheses I guess. It's pretty subjective, so debate on all points of the proposal is welcome and requested, even if you're just giving your opinion.</p>

<p>What follows is a proposal (that has a lot in common with the existing situation), then a summary of changes from the existing grammar, and then some examples.</p>

<h2 id="proposal">Proposal</h2>

<p>There are three types of operator:</p>

<ul>
<li>infix operators: always start with <code>x</code> and end with <code>y</code>.</li>

<li>prefix operators: always end with <code>x</code>.</li>

<li>suffix operators: always start with <code>x</code>.</li>

<li>primary expressions: not really operators, because they are either a single token, or they start and end with unambiguous grouping tokens.</li>
</ul>

<p>The tree below has ordered lists and unordered lists. The ordered lists specify a precedence relation in order of highest to lowest (1 is the highest precedence.). An unordered list specifies a group of items that cannot be intermixed with each other, unless the header for the list specifies otherwise.</p>

<p>Chainability comes in three forms:</p>

<ul>
<li>not chainable. There is no implicit associativity, so you must use parentheses to combine 3 or more operands. e.g. you can't do <code>a == b == c</code>.</li>

<li>chaininable somehow. This means for infix operators associativity is implicitly left-to-right when combining 3 or more operands with the operator in between them. e.g. <code>a + b + c</code> is implicitly equivalent to <code>(a + b) + c</code>. When a group (specified with an unordered list) of 2 or more items is chainiable, it will be one of these two subtypes:


<ul>
<li>heterogeneous chainable. This means all members of the group can be intermixed while chaining. e.g. <code>a + b - c</code>.</li>

<li>homogeneous chainable. Each member of the group can be chained only with itself. e.g. <code>a &amp; b &amp; c</code> is allowed, but <code>a &amp; b | c</code> is not.</li></ul>
</li>
</ul>

<ol>
<li>primary expressions, unambiguous:


<ul>
<li><code>123</code> - and variations, where <code>123</code> is a number literal</li>

<li><code>"..."</code> - where <code>...</code> is the characters of the string literal</li>

<li><code>'...'</code> - where <code>...</code> is the characters of the char literal</li>

<li><code>true</code>, <code>false</code>, <code>null</code>, <code>continue</code>, <code>undefined</code>, <code>error</code>, <code>this</code>, <code>unreachable</code></li>

<li><code>id</code> - where <code>id</code> is an identifier</li>

<li><code>@id(...)</code> - where <code>id</code> is an identifier and <code>...</code> is a parameter list</li>

<li><code>goto label</code> - where <code>label</code> is a label</li>

<li><code>break</code>, <code>return</code> - without an expression</li>

<li><code>(x)</code></li>

<li><code>{...}</code> - where <code>...</code> is a statement list</li>

<li><code>asm ("...")</code> - and variations, where <code>...</code> is inline assembly</li>

<li><code>struct { ... }</code> - and variations, where <code>...</code> is a container member list</li>

<li><code>if (a) {...}</code> - and variations, where <code>...</code> is a statement list[1]</li>

<li><code>if (a) b else {...}</code> - and variations, where <code>...</code> is a statement list[1]</li>

<li><code>while (a) {...}</code> - and variations, where <code>...</code> is a statement list[1]</li>

<li><code>while (a) b else {...}</code> - and variations, where <code>...</code> is a statement list[1]</li>

<li><code>for (a) {...}</code> - and variations, where <code>...</code> is a statement list[1]</li>

<li><code>for (a) b else {...}</code> - and variations, where <code>...</code> is a statement list</li>

<li><code>switch (a) {...}</code> - where <code>...</code> is a case list</li>

<li><code>comptime {...}</code> - where <code>...</code> is a statement list[1]</li></ul>
</li>

<li>suffix access and call operators, heterogeneous chainable:


<ul>
<li><code>x(...)</code> - where <code>...</code> is a parameter list</li>

<li><code>x[a]</code> - and variations</li>

<li><code>x.id</code> - where <code>id</code> is an identifier</li></ul>
</li>

<li>prefix high-precedent operators, heterogeneous chainable:


<ul>
<li><code>!x</code></li>

<li><code>-x</code>, <code>-%x</code></li>

<li><code>~x</code></li>

<li><code>*x</code></li>

<li><code>&amp;x</code> - and variations</li>

<li><code>??x</code></li>

<li><code>%%x</code></li>

<li><code>?x</code></li>

<li><code>%x</code></li>

<li><code>[]x</code> - and variations</li>

<li><code>extern fn() -&gt; x</code> - and variations </li></ul>
</li>

<li>suffix container initialization operator, not chainable:


<ul>
<li><code>x{...}</code> - where <code>...</code> is a container init body</li></ul>
</li>

<li>infix value manipulation operators:


<ul>
<li>math:


<ol>
<li>multiplicative:


<ul>
<li>scalar multiplicative, heterogeneous chainable:


<ul>
<li><code>x * y</code></li>

<li><code>x *% y</code></li>

<li><code>x / y</code></li></ul>
</li>

<li>modulo, not chainable:


<ul>
<li><code>x % y</code></li></ul>
</li>

<li>repetition, not chainable:


<ul>
<li><code>x ** y</code></li></ul>
</li></ul>
</li>

<li>additive:


<ul>
<li>scalar additive, heterogeneous chainable:


<ul>
<li><code>x + y</code></li>

<li><code>x +% y</code></li>

<li><code>x - y</code></li>

<li><code>x -% y</code></li></ul>
</li>

<li>concatenation, chainable:


<ul>
<li><code>x ++ y</code></li></ul>
</li></ul>
</li></ol>
</li>

<li>bit shifting, not chainable:


<ul>
<li><code>x &lt;&lt; y</code></li>

<li><code>x &lt;&lt;% y</code></li>

<li><code>x &gt;&gt; y</code></li></ul>
</li>

<li>bitwise algebra and monad coercion, homogeneous chainable:


<ul>
<li><code>x &amp; y</code></li>

<li><code>x ^ y</code></li>

<li><code>x | y</code></li>

<li><code>x ?? y</code></li>

<li><code>x %% y</code> - and variations</li></ul>
</li></ul>
</li>

<li>infix comparison operators, not chainable:


<ul>
<li><code>x == y</code></li>

<li><code>x != y</code></li>

<li><code>x &lt; y</code></li>

<li><code>x &lt;= y</code></li>

<li><code>x &gt; y</code></li>

<li><code>x &gt;= y</code></li></ul>
</li>

<li>infix boolean algebra operators, homogeneous chainable:


<ul>
<li><code>x and y</code></li>

<li><code>x or y</code></li></ul>
</li>

<li>prefix control operators, heterogeneous chainable:


<ul>
<li><code>return x</code></li>

<li><code>break x</code></li>

<li><code>if (a) x</code> - and variations[1]</li>

<li><code>if (a) b else x</code> - and variations[1]</li>

<li><code>while (a) x</code> - and variations[1]</li>

<li><code>while (a) b else x</code> - and variations[1]</li>

<li><code>for (a) x</code> - and variations[1]</li>

<li><code>for (a) b else x</code> - and variations[1]</li>

<li><code>comptime x</code>[1]</li></ul>
</li>
</ol>

<p>[1]: for the statement-like expressions <code>if</code>, <code>for</code>, <code>while</code>, and <code>comptime</code>, a special exception is made that if a body expression starts with a block (starts with a <code>{</code>), then the body expression ends at the end of that block (at the <code>}</code>). This disallows, for example, <code>if (a) {b} + 1 else {c}</code>.</p>

<h2 id="proposedchanges">Proposed Changes</h2>

<ul>
<li>Most of the "infix value manipulation operators" group is new.


<ul>
<li>Bitshifting cannot be chained.</li>

<li><code>++</code> cannot be mixed with <code>+</code> and the like.</li>

<li><code>**</code> cannot be mixed with <code>*</code> and the like, and cannot be chained.</li>

<li><code>%</code> cannot be mixed with other multiplicative operators, and is not chainable.</li>

<li>None of the bit manipulation operators can be intermixed with the math operators. Not even <code>a &lt;&lt; b + c</code>.</li>

<li><code>??</code> and <code>%%</code> infix operators cannot be intermixed.</li></ul>
</li>

<li><code>and</code> and <code>or</code> cannot be intermixed.</li>

<li>assignment is no longer an expression operator, but a statement construct. That's why it's not in the above tree.</li>

<li><code>inline</code> is gone; see #306.</li>
</ul>

<h2 id="examples">Examples</h2>

<ul>
<li>valid: <code>return a and b == c + d * e;</code></li>

<li>error: <code>x &lt;&lt; n + 1</code>. fixed: <code>x &lt;&lt; (n + 1)</code>.</li>

<li>error: <code>x &lt;&lt; 1 &lt;&lt; n</code>. fixed: <code>x &lt;&lt; (1 &lt;&lt; n)</code> (or whatever).</li>

<li>error: <code>" " ** spaces_per_indent ** indent_level</code>. fixed: <code>" " ** (spaces_per_indent * indent_level)</code>.</li>

<li>error: <code>foo() ?? bar() %% 0</code>. fixed: <code>foo() ?? (bar() %% 0)</code>.</li>
</ul>

<p>Here's some code that already follows this proposal from <code>std/rand.zig</code>:</p>

<ul>
<li><code>prev_value = int(i) +% f *% (prev_value ^ (prev_value &gt;&gt; (int.bit_count - 2)));</code></li>

<li><code>mt.array[i] = mt.array[i + m - n] ^ (x &gt;&gt; 1) ^ mag01[x &amp; 0x1];</code></li>
</ul>

<p>Here's some code from <code>std/base64.zig</code> that's fine:</p>

<ul>
<li><code>dest[out_index] = alphabet[((source[i] &amp; 0x3) &lt;&lt;% 4) | ((source[i + 1] &amp; 0xf0) &gt;&gt; 4)];</code></li>
</ul>

<p>But this code from <code>std/base64.zig</code> needs parentheses:</p>

<ul>
<li><code>dest[dest_index] = ascii6[source[src_index + 2]] &lt;&lt;% 6 | ascii6[source[src_index + 3]];</code></li>
</ul>

<p>Here's a curious case from <code>std/rand.zig</code>, where there are too many parentheses:</p>

<ul>
<li><code>return start + (rand_val % range);</code></li>
</ul>

<p>Perhaps that last case is a clue that we should not consider <code>%</code> to be a multiplicative math operator, but rather be a value manipulation operator like <code>&amp;</code> that doesn't mix with any other value manipulation operator.</p>

<p>I want <code>??</code> to be chainable because of this case:</p>

<ul>
<li>valid: <code>foo() ?? bar() ?? 0</code>. equivalent to <code>foo() ?? (bar() ?? 0)</code> and <code>(foo() ?? bar()) ?? 0</code>.</li>
</ul>

<p>But now that I think about that some more, it might not actually be correct. Consider if <code>foo()</code> returns a <code>??u32</code>. This means <code>(foo() ?? bar()) ?? 0</code> would unwrap both maybes, but <code>foo() ?? (bar() ?? 0)</code> would not. Hm. So either we need to define <code>??</code> as right-to-left associative, or make it not chainable. Similarly with <code>%%</code>.</p>

**Comment by andrewrk** [comment]
I agree with this proposal. Regarding the question at the end, let's just make it not chainable.

Regarding `%`, I think it's related enough to `/` that it should have the same chain properties as `/`.

**Comment by thejoshwolfe** [comment]
can you give a usecase for chaining `a % b % c`? it seems like you'd only ever want at most one `%` in a chained expression​.

**Comment by andrewrk** [comment]
I meant that you would be able to do return start + rand_val % range

**Comment by andrewrk** [comment]
I'll take a look at this before 0.2.0. If anyone wants to see this in 0.1.0 feel free to make a pull request.

**Comment by monouser7dig** [comment]
I made a proposal about the same issue here https://github.com/crystal-lang/crystal/issues/5074

based on this blog post https://foonathan.net/blog/2017/07/24/operator-precedence.html

with examples https://github.com/crystal-lang/crystal/issues/5074#issuecomment-352428851

It was generally well accepted but there was no one with sufficient expertise to make it happen

**Comment by lerno** [comment]
I would actually argue (see #1287) that bitshift operations should have higher precedence so that `a >> b + c  == (a >> b) + c` and also obviously allowing `b & 1 << 4 == b & (1 << 4)`

**Comment by markfirmware** [comment]
Too long won’t read what is going on?

**Comment by thejoshwolfe** [comment]
see the "Examples" in my long comment for proposed new errors.

**Comment by markfirmware** [comment]
@thejoshwolfe Thanks, I will look at that.

**Comment by tauoverpi** [comment]
Having `zig fmt` automatically fix such expressions with the correct paren nesting zig sees would be incredibly helpful as then  there's less of a need to place the parens manually and eliminates the possible learning curve to which expressions need them where. The writing case wouldn't be affected as much apart from having to remove extra parens which almost any editor handles with ease.

I wouldn't mind contributing it once I've gotten my head around the current formatting code and zig in general.

**Comment by Mouvedia** [comment]
Based on
> [I agree with this proposal.](https://github.com/ziglang/zig/issues/114#issuecomment-302245397)
I reckon there are 3 steps to this :

1. introduce the changes proposed by @thejoshwolfe before 1.0.0
2. have an experimental branch of fmt that automatically adds parentheses for a chosen subset of the operators
3. evaluate whether we keep it as is or make parentheses a requirement (hence move the precedence logic to fmt)

**Comment by displicit** [comment]
Completely unnecessary, it is a bit of the same as trying to enforce a markup style on programmer. Style of C preached by K&R is full of syntax that's hard to understand, yet laconic, unchainable expressions come here as a Rust-spirited addition.
Sure it forces programmers to write less error-prone code, but it makes the language more ugly - since it now has two types of expressions, chanable and unchainable - it is not worth it for this simple reason.

**Comment by ryancsh** [comment]
I think removing infix precedence would make things unambiguous, but also harder to read. 

By that I mean: the order in which things happen in would be clear, but would also slow down reading considerably. The human mind prefers infix over postfix and prefix, because infix is how natural languages (English, French, ...) are structured.

To illustrate:
```
infix:       2         +         3      
            boy       take      book 


prefix:      +         2         3
            take      boy       book 


postfix:     2         3         +
            boy       book      take 
```

A more complicated example:

```
infix:   2 + 3 + 4 + 5 + 6 + 7 + 8 + 9

prefix:  + + + + + + + 2 3 4 5 6 7 8 9

postfix: 2 3 4 5 6 7 8 9 + + + + + + +
```

With the prefix and postfix notation, I can't even visually make sure the number of operators is correct. Not only is it hard to write, but also hard to read.

-----

However, I do agree that the entire precedence table is difficult to remember and I find myself either looking it up or adding extra parentheses anyway to make things clear.

Thus, one radical way to solve this might be to entirely get rid of operator precedence for all binary operators (operators that take in two operands).

For example, the following two lines would be equivalent:

```
a * b + c / d % e % f + g << h & i | j and k or f

(((((((((((a * b) + c) / d) % e) % f) + g) << h) & i) | j) and k) or f)
```

Simply read from left to right. It's still somewhat awkward because of math classes hammering in the idea that multiplication happens before addition, but I think it's a reasonable compromise between readability and unambiguity. At the very least, I prefer this over the prefix / postfix notation.

However, getting rid of the entire precedence table and going left to right would lead to the following:

``` C
// consider this:
x += 3 * 5    

// it converts to
(x += 3) * 5

// to get original meaning we would need to do this
x += (3 * 5)
```

It's somewhat arguable if this really helps or not. Imagine reading through a 3-d math function full of this:

```
x += ( x - (y * z));
y += ( a + (b / 8));
z -= ( c | (d & e));
```

I feel like this might be more readable in some situations, but also involves more typing.

-----

**TLDR:**
1. Removing infix notation is not worth it. Slows down reading too much.
2. As an alternative, consider flattening all binary operators precedence, and instead having the leftmost operator acting first (assuming there are no parentheses).
3. I'm unsure about unary operators.

**Comment by Srekel** [comment]
A perhaps minor thing, but I feel like division should need parenthesis.

20 / 2 / 2, is it 1 or 5?
20 / 2 * 2, is it 20 or 5?

Maybe I should be 100% certain after programming for 20 years but... I'm not.

**Comment by ominitay** [comment]
I think it's fair enough to assume basic mathematical knowledge, with division and multiplication first, left-to-right, and then addition and subtraction, also left-to-right. So `20 / 2 / 2` is the same as `(20 / 2) / 2`, and `20 / 2 * 2` is `(20 / 2) * 2`

**Comment by exxjob** [comment]
Tracking:
https://github.com/ziglang/zig/issues/114
https://github.com/ziglang/zig/issues/14038
https://github.com/ziglang/zig/issues/1902
https://github.com/ziglang/zig/issues/15108
https://github.com/ziglang/zig/issues/10230

**Comment by exxjob** [comment]
Tracking:
* https://github.com/ziglang/zig/issues/114  | this
* https://github.com/ziglang/zig/issues/14038                                | subtask
* https://github.com/ziglang/zig/issues/1902                             | solution above
* https://github.com/ziglang/zig/issues/15108                                         | proposal
* https://github.com/ziglang/zig/issues/10230                               | docs
